eli5/keras.py

```diff
@@ -0,0 +1,293 @@
+# -*- coding: utf-8 -*-
+from __future__ import absolute_import
+
+import numpy as np
+import keras
+import keras.backend as K
+from keras.models import Model
+from keras.layers import Layer, Lambda
+from keras.preprocessing.image import load_img, img_to_array, array_to_img
+
+from eli5.base import Explanation
+from eli5.explain import explain_prediction
+
+
+DESCRIPTION_KERAS = """Grad-CAM visualization for image input; output is images"""
+
+# note that keras.models.Sequential subclasses keras.models.Model, so we can just register Model
+@explain_prediction.register(Model)
+def explain_prediction_keras(estimator, doc,
+                             target_names=None, # TODO: implement this
+                             targets=None,
+                             # new parameters:
+                             layer=None,
+                            ):
+    """
+    Explain an image prediction of a Keras image classifier.
+
+    See :func:`eli5.explain_prediction` for more information about the ``estimator``,
+    ``doc``, ``target_names``, and ``targets`` parameters.
+
+    Parameters
+    ----------
+    estimator : object
+        Instance of a Keras neural network model.
+    doc : object
+        An input image as a tensor to ``estimator``, for example ``numpy.ndarray``.
+    target_names : list, optional
+        *`Not Implemented`*.
+
+        Names for classes in the final output layer.
+    targets : list[int], optional
+        Prediction ID's to focus on.
+
+        *`Currently only the first prediction from the list is explained`*.
+
+        If None, the model is fed the input and its top prediction 
+        is taken as target as the target automatically.
+    layer : int or str or object, optional
+        The activation layer in the model to perform Grad-CAM on,
+        a valid keras layer name, layer index, or an instance of keras.layers.Layer.
+        
+        If None, a suitable layer is attempted to be retrieved (raise ValueError if can not).
+
+    Returns
+    -------
+    A :class:`base.Explanation` object with the ``image`` and ``heatmap`` attributes set.
+    """
+    activation_layer = get_activation_layer(estimator, layer)
+    predicted = get_target_prediction(estimator, doc, targets)
+    
+    heatmap = grad_cam(estimator, doc, predicted, activation_layer)
+    # TODO: consider renaming 'heatmap' to 'visualization'/'activations' (the output is not yet a heat map)
+
+    # need to insert a 'channel' axis to get a rank 3 image from rank 2
+    heatmap = np.expand_dims(heatmap, axis=-1)
+    heatmap = array_to_img(heatmap) # -> PIL image
```

Comment:
I'm curious why do we need to expand axis here and convert it to an image? To me a heatmap as a float numpy array in [0, 1] range would make more sense in the API, although I'm not 100% sure.

Reply:
I expanded the axis because `array_to_img()` (from keras preprocessing) requires a rank 3 array (the heatmap returned by grad cam is rank 2 / 2D).

Reply:
I think you are right. If the user accesses `.heatmap`, they can work with the values easily if the contents are a 2D numpy array.

One advantage of converting the heatmap to an image is that the user can readily see how the original heatmap looks. However, it's only a few lines to convert your suggested format into an image, so I think the user can do the conversion themselves.

Reply:
https://github.com/TeamHG-Memex/eli5/pull/315/commits/20d3635c8468f1702b619264f1296d4be1dbe283